backend/liveness/liveness.go

```diff
@@ -0,0 +1,298 @@
+
+// EtcdFactory returns a Factory that uses an etcd client. The Interface is
+// cached after the first instantiation, and the EventFuncs and logger cannot
+// be changed later.
+func EtcdFactory(ctx context.Context, client *clientv3.Client) Factory {
+	return Factory(func(name string, dead, alive EventFunc, logger logrus.FieldLogger) Interface {
+		switchMu.Lock()
+		defer switchMu.Unlock()
+		_, ok := switches[name]
+		if !ok {
+			ss := NewSwitchSet(client, name, dead, alive, logger)
+			ss.monitor(ctx)
+			switches[name] = ss
+		}
+		return switches[name]
+	})
+}
+
+// SwitchSet is a set of switches that get flipped on life and death events
+// for entities. On life and death events, callback functions that are
+// reigstered on NewSwitchSet are started as new goroutines.
+//
+// The SwitchSet uses the Alive method to both register members of the set,
+// and to assert their liveness once registered. After its first call to
+// Alive, if a member does not assert its liveness, then it will be presumed
+// to be dead, and the callback for dead members will be called.
+//
+// When an entity in a SwitchSet dies, it gains a new life in the underworld.
+// In the underworld, a dead callback is issued for every TTL interval.
+// Entities can go from being dead to alive by calling Alive. When that
+// happens, an entity that lives in the underworld will be reborn.
+type SwitchSet struct {
+	client      *clientv3.Client
+	prefix      string
+	notifyDead  EventFunc
+	notifyAlive EventFunc
+	logger      logrus.FieldLogger
+
+	// This channel serializes events so that their execution ordering is
+	// as expected, without causing undue blocking in the main monitoring
+	// loop.
+	events chan func()
+}
+
+// EventFunc is a function that can be used by a SwitchSet to handle events.
+// The previous state of the switch will be passed to the function.
+//
+// For "dead" EventFuncs, the leader flag can be used to determine if the
+// client that flipped the switch is our client. For "alive" EventFuncs,
+// this parameter is always false.
+type EventFunc func(key string, prev State, leader bool)
+
+// NewSwitchSet creates a new SwitchSet. It will use an etcd prefix of
+// path.Join(SwitchPrefix, name). The dead and live callbacks will be called
+// on all life and death events.
+func NewSwitchSet(client *clientv3.Client, name string, dead, alive EventFunc, logger logrus.FieldLogger) *SwitchSet {
+	return &SwitchSet{
+		client:      client,
+		prefix:      path.Join(SwitchPrefix, name),
+		notifyDead:  dead,
+		notifyAlive: alive,
+		logger:      logger,
+		events:      make(chan func(), 512),
+	}
+}
+
+func (t *SwitchSet) ping(ctx context.Context, key string, ttl int64, alive bool) error {
+	if ttl < FallbackTTL {
+		return fmt.Errorf("bad ttl: %d is less than the minimum value of %d", ttl, FallbackTTL)
+	}
+	if !alive {
+		ttl = -ttl
+	}
+	key = path.Join(t.prefix, key)
+	val := fmt.Sprintf("%x", ttl)
```

Comment:
I don't see a good rationale for hex other than that it's slightly more compact. Should I change it?

Reply:
I'm OK with keeping it in hex, as long as it's easy to tell if we go inspect the value in the store manually. Right now, and correct me if I'm wrong, you would store the decimal value `22` (for example) as the string `16`. I would argue it should be `0x16` so that I don't wrongly assume decimal 16 and get puzzled by whatever I'm trying to debug ;)

Reply:
Unfortunately I don't think -0x16 is a valid hex value in terms of Scanf.

Reply:
I'll just change it to decimal.

Reply:
You would have to do something like `Scanf("0x%x", &value)`. That should work. Changing to decimal also work. Your call!